### PR TITLE
[feat] Auction

### PR DESCRIPTION
**Pull request by vuvu2259**
- Add auction method into sdk

### CODE REVIEW

File: core/sdk/src/api/program/auction/create.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+export const createAuction = async (
+  wallet: AnchorWallet | Keypair,
+  treasuryMint: PublicKey,
+  nftMint: PublicKey,
+  startingBid: anchor.BN,
+  startTime: anchor.BN,
+  biddingPeriod: anchor.BN,
+  buyNowPrice: anchor.BN | null,
+  program: anchor.Program
```
Comment (haVincy) [inline core/sdk/src/api/program/auction/create.ts:14]: Can follow the structure in `core/sdk/src/api/model.ts` to organize the params for auction methods.

File: core/sdk/src/api/program/auction/bid.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+export const bid = async (
+  wallet: AnchorWallet | Keypair,
+  buyer: Keypair,
+  treasuryMint: PublicKey,
+  nftMint: PublicKey,
+  metadata: PublicKey,
+  auctionHouse: PublicKey,
+  feeAccount: PublicKey,
+  bidPrice: anchor.BN,
+  program: anchor.Program
```
Comment (haVincy) [inline core/sdk/src/api/program/auction/bid.ts:28]: ditto

File: core/sdk/src/api/utils.ts [hunk 3]
```diff
@@ -15,6 +25,13 @@ import { awaitTransactionSignatureConfirmation, WRAPPED_SOL_MINT } from '.';
 const METADATA_PROGRAM_ID = 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s';
 const metadataProgramId = new web3.PublicKey(METADATA_PROGRAM_ID);
 
+export const getAuction = async (candyShop: web3.PublicKey, mint: web3.PublicKey, marketProgramId: web3.PublicKey) => {
+  return web3.PublicKey.findProgramAddress(
+    [Buffer.from(AUCTION), candyShop.toBuffer(), mint.toBuffer()],
+    marketProgramId
+  );
+};
```
Comment (haVincy) [inline core/sdk/src/api/utils.ts:33]: There's no `await` so don't need `async` syntax, or we do following 

```ts
export const getAuction = async (candyShop: web3.PublicKey, mint: web3.PublicKey, marketProgramId: web3.PublicKey) => {
  const aution = await web3.PublicKey.findProgramAddress(
    [Buffer.from(AUCTION), candyShop.toBuffer(), mint.toBuffer()],
    marketProgramId;
  );
  return aution;
};
```

File: core/sdk/src/api/utils.ts [hunk 4]
```diff
@@ -268,3 +285,18 @@ export const sendTx = async (
...
+export const getBidWallet = async (
+  auction: web3.PublicKey,
+  wallet: web3.PublicKey,
+  marketProgramId: web3.PublicKey
+) => {
+  return web3.PublicKey.findProgramAddress(
+    [Buffer.from(BID), auction.toBuffer(), wallet.toBuffer(), Buffer.from(WALLET)],
+    marketProgramId
+  );
+};
```
Comment (haVincy) [inline core/sdk/src/api/utils.ts:302]: ditto. Either remove redundant `async` or we add `await`.

File: core/sdk/src/api/program/auction/bid.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  const [bidTradeState, bidTradeStateBump] = await getAuctionHouseTradeState(
+    auctionHouse,
+    bidWallet,
+    auctionEscrow,
+    treasuryMint,
+    nftMint,
+    new anchor.BN(1),
+    bidPrice
+  );
```
Comment (haVincy) [inline core/sdk/src/api/program/auction/bid.ts:58]: Theres's multiple `await` in one function call `bid`, I wonder if it's valid assumption that 

1. So if one of them is failed, we should not proceed, `bid` promise will throw error from it to caller.
2. Does `bid` has similar promises usage from `buyNow` ? If yes, we should organize them by extracting as reusable method for both `bid` and `buyNod` to use.

File: core/sdk/src/api/program/auction/requestExtraCompute.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import * as anchor from '@project-serum/anchor';
+import { TransactionInstruction } from '@solana/web3.js';
+import { COMPUTE_BUDGET_PROGRAM_ID } from '../../constants';
+
+export const requestExtraCompute = (amount: number) => {
```
Comment (haVincy) [inline core/sdk/src/api/program/auction/requestExtraCompute.ts:5]: `requestExtraComputeIx`

File: core/sdk/src/api/program/auction/bid.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  getBidWallet,
+  getCandyShop,
+  sendTx
+} from '../..';
+import { WRAPPED_SOL_MINT, AUCTION_HOUSE_PROGRAM_ID } from '../../constants';
+import { requestExtraCompute } from './requestExtraCompute';
+
+export const bid = async (
+  wallet: AnchorWallet | Keypair,
+  buyer: Keypair,
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/bid.ts:21]: The `buyer` is going to be the person calling the instruction. So type should be `AnchorWallet | Keypair`.

File: core/sdk/src/api/program/auction/bid.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  getBid,
+  getBidWallet,
+  getCandyShop,
+  sendTx
+} from '../..';
+import { WRAPPED_SOL_MINT, AUCTION_HOUSE_PROGRAM_ID } from '../../constants';
+import { requestExtraCompute } from './requestExtraCompute';
+
+export const bid = async (
+  wallet: AnchorWallet | Keypair,
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/bid.ts:20]: This should be the Candy Shop creator public key.

File: core/sdk/src/api/program/auction/cancel.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
+import * as anchor from '@project-serum/anchor';
+import { AnchorWallet } from '@solana/wallet-adapter-react';
+import { Keypair, PublicKey, SYSVAR_CLOCK_PUBKEY, Transaction } from '@solana/web3.js';
+import { getAtaForMint, getAuction, getAuctionHouseAuthority, getCandyShop, sendTx } from '../..';
+
+export const cancelAuction = async (
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/cancel.ts:6]: This instruction is only callable by the Candy Shop creator. I suggest that the passed `wallet` public key be verified against it. If not equal throw an error

File: core/sdk/src/api/program/auction/cancel.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  wallet: AnchorWallet | Keypair,
+  treasuryMint: PublicKey,
+  nftMint: PublicKey,
+  program: anchor.Program
+) => {
+  const [candyShop] = await getCandyShop(wallet.publicKey, treasuryMint, program.programId);
+
+  const [auction, auctionBump] = await getAuction(candyShop, nftMint, program.programId);
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/cancel.ts:14]: Add a preemptive check to ensure that the auction can be cancelled.

Something like:
```ts
const currentTime = Date.now / 1000;
const auctionData = await program.accounts.auctionV1.fetch(auction); // This line can throw an error. Maybe use safeAwait
const auctionEndTime = auctionData.startTime + auctionData.biddingPeriod;

if(currentTime > auctionEndTime && auctionData.highestBid != null){
    throw Error;
}

if(currentTime > auctionData.startTime && currentTime < auctionEndTime){
    throw Error;
}
```

File: core/sdk/src/api/program/auction/create.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
+import * as anchor from '@project-serum/anchor';
+import { AnchorWallet } from '@solana/wallet-adapter-react';
+import { Keypair, PublicKey, Transaction } from '@solana/web3.js';
+import { getAtaForMint, getAuction, getAuctionHouseAuthority, getCandyShop, sendTx } from '../..';
+
+export const createAuction = async (
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/create.ts:6]: This instruction is only callable by the Candy Shop creator. I suggest that the passed `wallet` public key be verified against it. If not equal throw an error

File: core/sdk/src/api/program/auction/bid.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    auctionEscrow,
+    treasuryMint,
+    nftMint,
+    new anchor.BN(1),
+    bidPrice
+  );
+
+  const transaction = new Transaction();
+
+  const ix = await program.methods
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/bid.ts:62]: I suggest you add some preemptive checks before this line to ensure the instruction will succeed.

Verify bidding period open:
```ts
const current_time = Date.now() / 1000;
const auctionData = await program.accounts.auctionV1.fetch(auction); // This line can throw an error. Maybe use safeAwait
const auctionEndTime = auctionData.startTime + auctionData.biddingPeriod;

if currentTime < auctionData.startTime {
    throw Error;
}
      
if currentTime > auctionEndTime {
    throw Error;
}
```

Verify bid price is high enough
```ts
const auctionData = await program.accounts.auctionV1.fetch(auction);

if(auctionData.highestBid && bidPrice < auctionData.highestBid) {
    throw Error;
} else if(!auctionData.highestBid && bidPrice < auctionData.startingBid) {
    throw Error;
}
```

File: core/sdk/src/api/program/auction/bid.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  wallet,
+  buyer,
+  treasuryMint,
+  nftMint,
+  metadata,
+  auctionHouse,
+  feeAccount,
+  bidPrice,
+  program
+}: BidAuctionParams) => {
```
Comment (haVincy) [inline core/sdk/src/api/program/auction/bid.ts:29]: Would like to have param be simply `params` in callee method as we already make an `interface` for it.

```ts
export const bidAuction = async (params: BidAuctionParams)) => {
    const wallet = params.wallet;
    const buyer = params.buyer;
    const treasuryMint = params.treasuryMint;
    ....
}
```

File: core/sdk/src/api/program/auction/buyNow.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  buyer,
+  treasuryMint,
+  nftMint,
+  metadata,
+  auctionHouse,
+  feeAccount,
+  treasuryAccount,
+  buyNowPrice,
+  program
+}: BuyNowAuctionParams) => {
```
Comment (haVincy) [inline core/sdk/src/api/program/auction/buyNow.ts:29]: ditto

File: core/ui/src/public/Auction/index.tsx [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+  walletConnectComponent: React.ReactElement;
+}
+
+enum StageEnum {
+  SELECT = 'SELECT',
+  FORM = 'FORM',
+  CONFIRM = 'CONFIRM'
+}
+
+export const Auction: React.FC<AuctionProps> = ({ candyShop, wallet, walletConnectComponent }) => {
```
Comment (haVincy) [inline core/ui/src/public/Auction/index.tsx:34]: Do we really need `CandyShop` Instance for Auction UI if we already decoupled the most of methods.

File: core/sdk/src/api/program/auction/buyNow.ts [hunk 16]
```diff
@@ -1 +1,133 @@
...
+  getAuctionHouseProgramAsSigner,
+  getAuctionHouseTradeState,
+  getCandyShop,
+  sendTx,
+  treasuryMintIsNative
+} from '../..';
+import { AUCTION_HOUSE_PROGRAM_ID } from '../../constants';
+import { BuyNowAuctionParams } from '../../model';
+
+export const buyNowAuction = async ({
```
Comment (nibbus0x) [inline core/sdk/src/api/program/auction/buyNow.ts:18]: Add preemtive checks to ensure
1) Buy now is enabled for the selected auction
2) The auction is still within the bidding period

File: core/sdk/src/api/program/auction/bid.ts [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+  const [auctionEscrow] = await getAtaForMint(nftMint, auction);
+
+  const [
+    [authority],
+    [bid],
+    [userTreasuryAta],
+    [escrowPaymentAccount, escrowPaymentAccountBump],
+    [bidTradeState, bidTradeStateBump]
+  ] = await Promise.all([
```
Comment (RustySol) [inline core/sdk/src/api/program/auction/bid.ts:48]: don't think we should use `Promise.all()` here as all the functions inside are all non-async

File: core/sdk/src/CandyShop.ts [hunk 18]
```diff
@@ -340,6 +361,302 @@ export class CandyShop {
...
+   * @param {CandyShopCancelAuctionParams} params required parameters for sell action
+   */
+  public async cancelAuction(params: CandyShopCancelAuctionParams): Promise<string> {
+    const { tokenAccount, tokenMint, wallet } = params;
+
+    if (wallet.publicKey.toString() !== this.candyShopCreatorAddress.toString()) {
+      throw new CandyShopError(CandyShopErrorType.NonShopOwner);
+    }
+
+    console.log('CandyShop: performing cancel auction', {
```
Comment (haVincy) [inline core/sdk/src/CandyShop.ts:430]: We can organize the log by following before we have global Logger object.

```ts
const Logger = 'CandyShop';

....
console.log(`${Logger}: performing cancel auction`, {...});
```

File: core/sdk/src/api/utils.ts [hunk 19]
```diff
@@ -219,6 +250,108 @@ export const checkDelegateOnReceiptAccounts = async (
   }
 };
 
+export const getAuctionData = async (auction: web3.PublicKey, program: Program) => {
+  const auctionData = (await safeAwait(program.account.auctionV1.fetch(auction))).result;
+
+  if (!auctionData) {
+    throw new CandyShopError(CandyShopErrorType.AuctionDoesNotExist);
+  }
```
Comment (haVincy) [inline core/sdk/src/api/utils.ts:258]: We should also emit error while `safeAwait` resolves with error here. So the subscriber with this Promise call end up will catch the error in .catch, and the action will not proceed.

```ts
const auctionData = await safeAwait(program.account.auctionV1.fetch(auction));
if (auctionData.error) {
   throw new Error(auctionData.error);
}
if (!auctionData.result) {
    throw new CandyShopError(CandyShopErrorType.AuctionDoesNotExist);
}

return auctionData.result;
```

File: core/sdk/src/api/utils.ts [hunk 20]
```diff
@@ -219,6 +250,108 @@ export const checkDelegateOnReceiptAccounts = async (
...
+};
+
+export const getBidData = async (bid: web3.PublicKey, program: Program) => {
+  const bidData = (await safeAwait(program.account.bid.fetch(bid))).result;
+
+  if (!bidData) {
+    throw new CandyShopError(CandyShopErrorType.BidDoesNotExist);
+  }
+
+  return bidData;
```
Comment (haVincy) [inline core/sdk/src/api/utils.ts:270]: ditto.

**Comment by haVincy** [comment]
Test link is alive, any new push in this PR will auto deploy to AWS by Amplify.

- https://feat-auction.d22tqykss69onq.amplifyapp.com/

File: core/ui/src/components/AuctionModal/AuctionModalDetail.tsx [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+  const [countdown] = useState<number>(() => {
+    const second = Math.floor(Number(order.startTime) + Number(order.biddingPeriod) - dayjs().unix());
+    return dayjs().add(second, 'second').unix();
+  });
+  const [countDownRef, setCountDownRef] = useState<HTMLSpanElement | null>();
+  const [price, setPrice] = useState<number>();
+
+  const timeoutRef = useUnmountTimeout();
+
+  // useEffect(() => {
```
Comment (vanminh1701) [inline core/ui/src/components/AuctionModal/AuctionModalDetail.tsx:68]: remove comment

File: core/ui/src/public/Auctions/index.tsx [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+      limit: ORDER_FETCH_LIMIT,
+      status: [AuctionStatus.CREATED, AuctionStatus.EXPIRED, AuctionStatus.STARTED]
+    })
+      .then((data: any) => {
+        if (!data.result) return;
+        if (data.offset + data.count >= data.totalCount) {
+          setHasNextPage(false);
+        } else {
+          setHasNextPage(true);
+        }
```
Comment (vanminh1701) [inline core/ui/src/public/Auctions/index.tsx:50]: we just need: 
```
setHasNextPage(data.offset + data.count < data.totalCount)
```

File: core/ui/src/public/Auctions/index.tsx [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+
+  useEffect(() => {
+    fetchAuctionsByShopAddress(candyShop.candyShopAddress.toString(), {
+      offset: 0,
+      limit: ORDER_FETCH_LIMIT,
+      status: [AuctionStatus.CREATED, AuctionStatus.EXPIRED, AuctionStatus.STARTED]
+    })
+      .then((data: any) => {
+        if (!data.result) return;
+        const haveNextPage = data.offset + data.count < data.totalCount;
```
Comment (vanminh1701) [inline core/ui/src/public/Auctions/index.tsx:71]: don't need to create new variable `haveNextPage`

File: core/ui/src/public/Auctions/index.tsx [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+
+                const hoursString = `0${hours}`.slice(-2);
+                const minutesString = `0${minutes}`.slice(-2);
+                const secondsString = `0${seconds}`.slice(-2);
+
+                countdownElement.textContent = `Ends In: ${hoursString}:${minutesString}:${secondsString}`;
+
+                setTimeout(getRetainCountdown, 1000);
+              };
+              getRetainCountdown();
```
Comment (vanminh1701) [inline core/ui/src/public/Auctions/index.tsx:138]: please move ` getRetainCountdown` to useEffect hook, so it won't be re-initialize every render.
`
useEffect(() => {
  if(!nfts?.length) return;
  const getRetainCountdown = () => {
    nfts.forEach(item => {
      // logic update for each auction
     timeoutRef.current = setTimeout(...)
    })
  getRetainCountdown()
  return () => clearTimeout(timeoutRef.current)
}
},[nfts])
`

File: core/ui/src/public/Auctions/index.tsx [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+                <Card
+                  onClick={() => setSelected(order)}
+                  key={order.auctionAddress}
+                  name={order.name}
+                  ticker={order.symbol}
+                  imgUrl={order.image || ''}
+                  label={
+                    second <= 0 ? (
+                      <div className="candy-status-tag candy-status-tag-gray-1">ENDED</div>
+                    ) : order.highestBidBuyer === wallet?.publicKey.toString() ? (
```
Comment (vanminh1701) [inline core/ui/src/public/Auctions/index.tsx:150]: create `const walletKeyString = wallet?.publicKey.toString()` above return so we don't need to run this function for each auction card

File: core/ui/src/public/Auctions/index.tsx [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+                      <div className="candy-status-tag">HIGHEST BID</div>
+                    ) : (
+                      <div className="candy-status-tag candy-status-tag-yellow">OUTBID</div>
+                    )
+                  }
+                  footer={
+                    <div className="candy-card-footer">
+                      <div className="candy-card-stat">Current Bid: {highestBidPriceContent}</div>
+                      <div className="candy-card-stat" id={'x' + order.auctionAddress}>
```
Comment (vanminh1701) [inline core/ui/src/public/Auctions/index.tsx:159]: why we need plus `x` to `id`

**Comment by RustySol** [comment]
closed for #191

**Pull request closed without merging**